File: linkedList/cppRealizations/removeNthNodeFromEndofList.cpp
#include <iostream>

using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* left = dummy;
        ListNode* right = head;

        while(n > 0){
            right = right->next;
            n--;
        }

        while(right != nullptr){
            right = right->next;
            left = left->next;
        }

        // delete
        left->next = left->next->next;
        return dummy->next;
    }
};


int main(){
    ListNode *list = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    list = Solution().removeNthFromEnd(list, 2);
    while(list){
        cout << list->val << endl;
        list = list->next;
    }
    return 0;
}

File: linkedList/removeNthNodeFromEndofList.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def removeNthFromEnd(self, head, n):
        dummy = ListNode(0, head)
        left = dummy
        right = head

        while n > 0:
            right = right.next
            n -= 1

        while right:
            right = right.next
            left = left.next

        # delete
        left.next = left.next.next
        return dummy.next


if __name__ == '__main__':
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    temp = head
    print("Input:")
    while temp:
        print(temp.val)
        temp = temp.next
    print("Solution:")
    head = Solution().removeNthFromEnd(head, 2)
    while head:
        print(head.val)
        head = head.next
        